payjoin/src/send/mod.rs

```diff
@@ -239,56 +271,81 @@ impl Configuration {
-
-    /// Disable output substitution even if the receiver didn't.
-    ///
-    /// This forbids receiver switching output or decreasing amount.
-    /// It is generally **not** recommended to set this as it may prevent the receiver from
-    /// doing advanced operations such as opening LN channels and it also guarantees the
-    /// receiver will **not** reward the sender with a discount.
-    pub fn always_disable_output_substitution(mut self, disable: bool) -> Self {
-        self.disable_output_substitution = disable;
-        self
-    }
-
-    /// Decrease fee contribution instead of erroring.
-    ///
-    /// If this option is set and a transaction with change amount lower than fee
-    /// contribution is provided then instead of returning error the fee contribution will
-    /// be just lowered to match the change amount.
-    pub fn clamp_fee_contribution(mut self, clamp: bool) -> Self {
-        self.clamp_fee_contribution = clamp;
-        self
+    pub fn build_non_incentivizing(mut self) -> Result<(Request, Context), CreateRequestError> {
+        // since this is a builder, these should already be cleared
+        // but we'll reset them to be sure
+        self.fee_contribution = None;
+        self.clamp_fee_contribution = false;
+        self.min_fee_rate = FeeRate::ZERO;
+        self.build()
     }
 
-    /// Sets minimum fee rate required by the sender.
-    pub fn min_fee_rate(mut self, fee_rate: FeeRate) -> Self {
-        self.min_fee_rate = fee_rate;
-        self
+    fn build(self) -> Result<(Request, Context), CreateRequestError> {
+        let mut psbt =
+            self.psbt.validate().map_err(InternalCreateRequestError::InconsistentOriginalPsbt)?;
+        psbt.validate_input_utxos(true)
+            .map_err(InternalCreateRequestError::InvalidOriginalInput)?;
+        let disable_output_substitution =
+            self.uri.extras.disable_output_substitution || self.disable_output_substitution;
+        let payee = self.uri.address.script_pubkey();
+
+        check_single_payee(&psbt, &payee, self.uri.amount)?;
+        let fee_contribution = determine_fee_contribution(
+            &psbt,
+            &payee,
+            self.fee_contribution,
+            self.clamp_fee_contribution,
+        )?;
+        clear_unneeded_fields(&mut psbt);
+
+        let zeroth_input = psbt.input_pairs().next().ok_or(InternalCreateRequestError::NoInputs)?;
+
+        let sequence = zeroth_input.txin.sequence;
+        let txout = zeroth_input.previous_txout().expect("We already checked this above");
+        let input_type = InputType::from_spent_input(txout, zeroth_input.psbtin).unwrap();
+        let url = serialize_url(
+            self.uri.extras._endpoint.into(),
+            disable_output_substitution,
+            fee_contribution,
+            self.min_fee_rate,
+        )
+        .map_err(InternalCreateRequestError::Url)?;
+        let body = serialize_psbt(&psbt);
+        Ok((
+            Request { url, body },
```

Comment:
can we return the url as part of the context and then leave it to the receiver to construct the `Request` as they already have the `PSBT`?
that will allow us to drop the request from the return here I guess..

Reply:
The point of this interface is to create the request on the behalf of the client in such a way that they make as few decisions as possible and have as few opportunities for errors as possible. `Request` is thus a complete, serialized HTTP request that gets consumed by the http client, while the `Context` sticks around and is only the data required to handle the response.

Because Request memory gets moved we want these to be distinct structs as detailed in [this comment](https://github.com/payjoin/rust-payjoin/pull/87#issuecomment-1664721057)